src/coreclr/jit/runtimelookup.cpp

```diff
@@ -0,0 +1,461 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#include "jitpch.h"
+#ifdef _MSC_VER
+#pragma hdrstop
+#endif
+
+//------------------------------------------------------------------------------
+// fgExpandRuntimeLookups : partially expand runtime lookups helper calls
+//                          to add a nullcheck [+ size check] and a fast path
+// Returns:
+//    PhaseStatus indicating what, if anything, was changed.
+//
+PhaseStatus Compiler::fgExpandRuntimeLookups()
+{
+    PhaseStatus result = PhaseStatus::MODIFIED_NOTHING;
+
+    if (!doesMethodHaveExpRuntimeLookup())
+    {
+
+#ifdef DEBUG
+        for (BasicBlock* block : Blocks())
+        {
+            for (Statement* stmt : block->Statements())
+            {
+                for (GenTree* tree : stmt->TreeList())
+                {
+                    assert(!tree->IsCall() || (tree->IsCall() && !tree->AsCall()->IsExpRuntimeLookup()));
+                }
+            }
+        }
+#endif
+        JITDUMP("Current method doesn't have runtime lookups - bail out.")
+        return result;
+    }
+
+    // Find all calls with GTF_CALL_M_EXP_RUNTIME_LOOKUP flag
+    for (BasicBlock* block : Blocks())
+    {
+    TRAVERSE_BLOCK_AGAIN:
+
+        Statement* prevStmt = nullptr;
+        for (Statement* const stmt : block->Statements())
+        {
+            for (GenTree* const tree : stmt->TreeList())
+            {
+                // We only need calls with IsExpRuntimeLookup() flag
+                if (!tree->IsCall() || !tree->AsCall()->IsExpRuntimeLookup())
+                {
+                    continue;
+                }
+                assert(tree->IsHelperCall());
+
+                JITDUMP("Expanding runtime lookup for [%06d] in " FMT_BB ":\n", dspTreeID(tree), block->bbNum)
+                DISPTREE(tree)
+                JITDUMP("\n")
+
+                GenTreeCall* call = tree->AsCall();
+                call->ClearExpRuntimeLookup();
+                assert(call->gtArgs.CountArgs() == 2);
+
+                if (call->IsTailCall())
+                {
+                    assert(!"Unexpected runtime lookup as a tail call");
```

Comment:
Is this situation impossible? What about a function that e.g. returns `ldtoken` of a type argument?